File: mp2i/files/C7/wolfram.c
#include <stdbool.h>
#include <stdlib.h>
#include <stdio.h>

struct config_s
{
    int size;
    bool *cells;
};
typedef struct config_s config;

config init(int size, bool state)
{
    config c;
    c.size = size;
    c.cells = malloc(sizeof(bool) * size);
    for (int i = 0; i < size; i++)
    {
        c.cells[i] = state;
    }
    return c;
}

bool rule90(bool prev, bool current, bool next)
{
    if ((!prev && !current && !next) || (!prev && current && !next) || (prev && !current && next) || (prev && current && next))
    {
        return false;
    }
    return true;
}

bool rule(bool prev, bool current, bool next, int rnumber)
{
    int cvalue = 0;
    if (next)
    {
        cvalue += 1;
    }
    if (current)
    {
        cvalue += 2;
    }
    if (prev)
    {
        cvalue += 4;
    }
    return ((rnumber >> cvalue & 1) == 1);
}

void evolution(config *c)
{
    bool prev, current, next;
    bool *newcells = malloc(sizeof(bool) * c->size);
    for (int i = 0; i < c->size; i++)
    {
        current = c->cells[i];
        prev = i == 0 ? false : c->cells[i - 1];
        next = i == c->size - 1 ? false : c->cells[i + 1];
        newcells[i] = rule(prev, current, next, 60);
    }
    free(c->cells);
    c->cells = newcells;
}

int alive(config c)
{
    int a = 0;
    for (int i = 0; i < c.size; i++)
    {
        if (c.cells[i])
        {
            a++;
        }
    }
    return a;
}

void display(config c)
{
    for (int i = 0; i < c.size; i++)
    {
        if (c.cells[i])
        {
            printf("#");
        }
        else
        {
            printf(".");
        }
    }
    printf("\n");
}

void destroy(config *c)
{
    c->size = 0;
    free(c->cells);
}

int main()
{
    config c = init(500, false);
    c.cells[1] = true;
    for (int i = 0; i < 500; i++)
    {
        //display(c);
        evolution(&c);
    }
    printf("Résultat = %d \n", alive(c));
    destroy(&c);
    return 0;
}

File: mp2i/files/C7/wolfram.ml


let init n b =
  List.init n (fun i->b);;

let display c =
  List.iter  (fun b -> if b then print_char '#' else print_char '.') c;
  print_newline();;

let rec set_cell config idx b = 
  match config with
  | [] -> failwith "Index hors de la liste"
  | h::t -> if idx=0 then b::t else h::(set_cell t (idx-1) b);;

let rule90 prev current next =
  match prev, current, next with
  | false, false, false -> false
  | false, true, false -> false
  | true, false, true -> false
  | true, true, true -> false
  | _, _, _ -> true;;


let evolution config =
  let rec aux cf prev newc =
    match cf with
    | [] -> [];
    | h::[] -> (rule90 prev h false)::newc
    | h1::h2::t -> aux (h2::t) h1 ((rule90 prev h1 h2)::newc) 
  in
  List.rev (aux config false []);;



let () =
 let c = init 20 false in
 let c = set_cell c 10 true in
 display c;
 let c = evolution c in
 display c;
 let c = evolution c in
 display c;
 let c = evolution c in
 display c;
 let c = evolution c in
 display c;
 let c = evolution c in
 display c;
 let c = evolution c in
 display c;;
